File: clients/unreal/Source/FomoxaExample/Network/FomoxaExampleSession.cpp
#include "Network/FomoxaExampleSession.h"

#include <algorithm>
#include <cstdio>
#include <utility>

#include "Generated/handshake.hpp"
#include "Network/FomoxaExampleProtocol.h"

namespace FomoxaExample
{

Session::~Session()
{
    Close();
}

void Session::Open(const std::string& Host, uint16_t Port, const std::string& PlayerName)
{
    Close();
    FailureReason.clear();

    const std::string Address = Host + ":" + std::to_string(Port);
    std::optional<fomoxa::Transport> Transport = fomoxa::Transport::tcp(Host, Port);
    if (!Transport)
    {
        Fail("cannot reach " + Address);
        return;
    }

    Connection = fomoxa::Connection::create(std::move(*Transport), Schema());
    if (!Connection)
    {
        Fail("cannot start a session with " + Address);
        return;
    }

    DisplayName = PlayerName;
    CurrentStatus = SessionStatus::Connecting;
}

void Session::Close()
{
    Release();
    CurrentStatus = SessionStatus::Disconnected;
}

void Session::Poll(uint64_t NowMs)
{
    if (!Connection)
    {
        return;
    }

    for (fomoxa::Event Raised : Connection->tick(NowMs))
    {
        switch (Raised.kind())
        {
        case FMX_EVENT_READY:
            CurrentStatus = SessionStatus::Joining;
            break;
        case FMX_EVENT_MESSAGE:
            HandleMessage(Raised.message_id(), Raised.payload());
            break;
        case FMX_EVENT_HANDSHAKE_FAILED:
            Fail(std::string("handshake refused: ") + fmx_handshake_failure_name(Raised.handshake_failure()));
            return;
        case FMX_EVENT_DISCONNECTED:
            Fail(std::string("disconnected: ") + fmx_disconnect_name(Raised.disconnect_reason()));
            return;
        default:
            break;
        }
    }

    if (CurrentStatus == SessionStatus::Joining && !bHelloSent)
    {
        const std::vector<uint8_t> Payload = EncodeHello(DisplayName);
        bHelloSent = Connection->send(generated::CLIENT_HELLO_GAME_MESSAGE_ID, fomoxa::ByteView(Payload)) == FMX_OK;
    }
}

void Session::UpdateInput(uint64_t NowMs, float MoveX, float MoveZ, bool bJump, float LookYaw, float LookPitch)
{
    if (CurrentStatus != SessionStatus::Joined || NowMs < NextInputAtMs)
    {
        return;
    }

    FomoxaExampleModels::PlayerInput Input;
    Input.Sequence = ++InputSequence;
    Input.MoveX = MoveX;
    Input.MoveZ = MoveZ;
    Input.Jump = bJump;
    Input.LookYaw = LookYaw;
    Input.LookPitch = LookPitch;

    const std::vector<uint8_t> Payload = EncodeInput(Input);
    if (Connection->send(generated::PLAYER_INPUT_GAME_MESSAGE_ID, fomoxa::ByteView(Payload)) == FMX_OK)
    {
        NextInputAtMs = NowMs + 1000u / std::max<uint32_t>(TickRate, 1u);
    }
}

const FomoxaExampleModels::PlayerState* Session::FindLocalState() const
{
    const auto Found = Players.find(LocalPlayerId);
    return Found == Players.end() ? nullptr : &Found->second;
}

void Session::HandleMessage(uint32_t MessageId, fomoxa::ByteView Payload)
{
    std::string Error;
    if (MessageId == generated::WELCOME_GAME_MESSAGE_ID)
    {
        FomoxaExampleModels::Welcome Welcome;
        if (DecodeWelcome(Payload, Welcome, Error))
        {
            LocalPlayerId = Welcome.PlayerId;
            PlaneHalfSize = Welcome.PlaneHalfSize;
            TickRate = Welcome.TickRate;
            CurrentStatus = SessionStatus::Joined;
        }
    }
    else if (MessageId == generated::WORLD_SNAPSHOT_GAME_MESSAGE_ID)
    {
        FomoxaExampleModels::WorldSnapshot Snapshot;
        if (DecodeSnapshot(Payload, Snapshot, Error))
        {
            LastTick = Snapshot.Tick;
            ++SnapshotsReceived;
            Players.clear();
            for (FomoxaExampleModels::PlayerState& State : Snapshot.Players)
            {
                const uint32_t PlayerId = State.PlayerId;
                Players[PlayerId] = std::move(State);
            }
        }
    }

    if (!Error.empty())
    {
        char MessageLabel[16];
        std::snprintf(MessageLabel, sizeof(MessageLabel), "0x%08X", static_cast<unsigned>(MessageId));
        LastDecodeError = std::string(MessageLabel) + ": " + Error;
    }
}

void Session::Fail(const std::string& Reason)
{
    Release();
    FailureReason = Reason;
    CurrentStatus = SessionStatus::Failed;
}

void Session::Release()
{
    if (Connection)
    {
        Connection->close();
        Connection.reset();
    }
    bHelloSent = false;
    InputSequence = 0;
    NextInputAtMs = 0;
    LocalPlayerId = 0;
    LastTick = -1;
    SnapshotsReceived = 0;
    Players.clear();
}

}

File: clients/unity/Assets/FomoxaExample/Networking/FomoxaExampleSession.cs
using System;
using System.Collections.Generic;
using FomoxaExample.Networking.Models;
using Fomoxa.Net;
using Fomoxa.Net.Transports;
using Generated;

namespace FomoxaExample.Networking
{
    public enum FomoxaExampleStatus
    {
        Disconnected,
        Connecting,
        Joining,
        Joined,
        Failed,
    }

    public sealed class FomoxaExampleSession : IDisposable
    {
        readonly Dictionary<uint, PlayerState> players = new Dictionary<uint, PlayerState>();
        FomoxaConnection connection;
        string displayName = "";
        bool helloSent;
        uint inputSequence;
        TimeSpan nextInputAt;

        public FomoxaExampleStatus Status { get; private set; } = FomoxaExampleStatus.Disconnected;
        public string FailureReason { get; private set; } = "";
        public string LastDecodeError { get; private set; } = "";
        public uint LocalPlayerId { get; private set; }
        public float PlaneHalfSize { get; private set; } = 10f;
        public int TickRate { get; private set; } = 30;
        public long LastTick { get; private set; } = -1;
        public int SnapshotsReceived { get; private set; }
        public IReadOnlyDictionary<uint, PlayerState> Players => players;

        public bool TryGetLocalState(out PlayerState state) => players.TryGetValue(LocalPlayerId, out state);

        public void Open(string host, int port, string playerName)
        {
            Close();
            FailureReason = "";
            try
            {
                var transport = TcpTransport.Connect(host, port);
                connection = FomoxaConnection.Connect(transport, FomoxaExampleProtocol.BuildSchema(), new SessionConfig());
            }
            catch (Exception exception)
            {
                Fail($"cannot reach {host}:{port} - {exception.Message}");
                return;
            }
            displayName = playerName;
            Status = FomoxaExampleStatus.Connecting;
        }

        public void Close()
        {
            Release();
            Status = FomoxaExampleStatus.Disconnected;
        }

        public void Dispose() => Close();

        public void Poll() => Poll(MonotonicClock.Now);

        public void Poll(TimeSpan now)
        {
            if (connection == null)
            {
                return;
            }

            var events = connection.Tick(now);
            for (var index = 0; index < events.Count; index++)
            {
                var raised = events[index];
                switch (raised.Kind)
                {
                    case FomoxaEventKind.Ready:
                        Status = FomoxaExampleStatus.Joining;
                        break;
                    case FomoxaEventKind.Message:
                        HandleMessage(raised.MessageId, raised.Payload.Span);
                        break;
                    case FomoxaEventKind.HandshakeFailed:
                        Fail($"handshake refused: {raised.Failure}");
                        return;
                    case FomoxaEventKind.Disconnected:
                        Fail($"disconnected: {raised.Reason}");
                        return;
                }
            }

            if (Status == FomoxaExampleStatus.Joining && !helloSent)
            {
                helloSent = connection.Send(ClientHelloGameCodec.MessageId, FomoxaExampleProtocol.EncodeHello(displayName)) == SendStatus.Sent;
            }
        }

        public void UpdateInput(float moveX, float moveZ, bool jump, float lookYaw, float lookPitch) =>
            UpdateInput(MonotonicClock.Now, moveX, moveZ, jump, lookYaw, lookPitch);

        public void UpdateInput(TimeSpan now, float moveX, float moveZ, bool jump, float lookYaw, float lookPitch)
        {
            if (Status != FomoxaExampleStatus.Joined || now < nextInputAt)
            {
                return;
            }

            inputSequence++;
            var input = new PlayerInput
            {
                Sequence = inputSequence,
                MoveX = moveX,
                MoveZ = moveZ,
                Jump = jump,
                LookYaw = lookYaw,
                LookPitch = lookPitch,
            };
            if (connection.Send(PlayerInputGameCodec.MessageId, FomoxaExampleProtocol.EncodeInput(input)) == SendStatus.Sent)
            {
                nextInputAt = now + TimeSpan.FromSeconds(1.0 / Math.Max(TickRate, 1));
            }
        }

        void HandleMessage(uint messageId, ReadOnlySpan<byte> payload)
        {
            try
            {
                if (messageId == WelcomeGameCodec.MessageId)
                {
                    var welcome = FomoxaExampleProtocol.DecodeWelcome(payload);
                    LocalPlayerId = welcome.PlayerId;
                    PlaneHalfSize = welcome.PlaneHalfSize;
                    TickRate = welcome.TickRate;
                    Status = FomoxaExampleStatus.Joined;
                }
                else if (messageId == WorldSnapshotGameCodec.MessageId)
                {
                    var snapshot = FomoxaExampleProtocol.DecodeSnapshot(payload);
                    LastTick = snapshot.Tick;
                    SnapshotsReceived++;
                    players.Clear();
                    foreach (var state in snapshot.Players)
                    {
                        players[state.PlayerId] = state;
                    }
                }
            }
            catch (DecodeException exception)
            {
                LastDecodeError = $"0x{messageId:X8}: {exception.Message}";
            }
        }

        void Fail(string reason)
        {
            Release();
            FailureReason = reason;
            Status = FomoxaExampleStatus.Failed;
        }

        void Release()
        {
            if (connection != null)
            {
                connection.Close();
                connection.Dispose();
                connection = null;
            }
            helloSent = false;
            inputSequence = 0;
            nextInputAt = TimeSpan.Zero;
            LocalPlayerId = 0;
            LastTick = -1;
            SnapshotsReceived = 0;
            players.Clear();
        }
    }
}
